File: src/components/lib/message-adapter.ts
/**
 * Message adapter for TKR Multi-Agent Chat System.
 * 
 * This module provides adapters for converting between standardized and legacy message formats
 * to ensure compatibility during the transition period.
 */

import { 
  Message, 
  MessageType, 
  BaseMessage, 
  UserTextMessage, 
  AgentTextMessage,
  AgentToAgentMessage,
  SystemMessage,
  ErrorMessage,
  ContextUpdateMessage,
  TaskUpdateMessage,
  PingMessage,
  PongMessage,
  MessageStatus,
  MessageAcknowledgment
} from '../../types/messages';

/**
 * Convert a legacy format message to standardized format.
 * 
 * @param legacyMessage The legacy message object
 * @returns Standardized message object
 */
export function legacyToStandard(legacyMessage: any): Partial<Message> {
  // Base message fields
  const standardMessage: Partial<BaseMessage> = {
    id: legacyMessage.id || `msg_${Date.now()}`,
    timestamp: legacyMessage.timestamp || new Date().toISOString(),
    session_id: legacyMessage.sessionId || legacyMessage.session_id || 'default',
  };
  
  // Determine message type
  const type = legacyMessage.type || 'text';
  standardMessage.type = type as MessageType;
  
  // Add content - both content and text are used in legacy messages
  if ('content' in legacyMessage) {
    standardMessage.content = legacyMessage.content;
  } else if ('text' in legacyMessage) {
    standardMessage.content = legacyMessage.text;
  }
  
  // Handle agent vs user specifics
  if (legacyMessage.fromAgent || legacyMessage.from_agent) {
    standardMessage.from_agent = legacyMessage.fromAgent || legacyMessage.from_agent;
  } else if (legacyMessage.fromUser || legacyMessage.from_user) {
    standardMessage.from_user = true;
  } else {
    // Default to user message if no source specified
    standardMessage.from_user = true;
  }
  
  // Handle target agent
  if (legacyMessage.toAgent || legacyMessage.to_agent) {
    standardMessage.to_agent = legacyMessage.toAgent || legacyMessage.to_agent;
  }
  
  // Copy additional fields if present
  if ('inReplyTo' in legacyMessage || 'in_reply_to' in legacyMessage) {
    standardMessage.in_reply_to = legacyMessage.inReplyTo || legacyMessage.in_reply_to;
  }
  
  if ('streaming' in legacyMessage) {
    standardMessage.streaming = legacyMessage.streaming;
  }
  
  if ('turnComplete' in legacyMessage || 'turn_complete' in legacyMessage) {
    standardMessage.turn_complete = legacyMessage.turnComplete || legacyMessage.turn_complete;
  }
  
  if ('metadata' in legacyMessage) {
    standardMessage.metadata = legacyMessage.metadata;
  }
  
  // Handle message type specific fields
  switch (type) {
    case 'context_update':
      (standardMessage as Partial<ContextUpdateMessage>).context_id = 
        legacyMessage.contextId || legacyMessage.context_id;
      (standardMessage as Partial<ContextUpdateMessage>).context_data = 
        legacyMessage.contextData || legacyMessage.context_data;
      if ('targetAgents' in legacyMessage || 'target_agents' in legacyMessage) {
        (standardMessage as Partial<ContextUpdateMessage>).target_agents = 
          legacyMessage.targetAgents || legacyMessage.target_agents;
      }
      break;
    
    case 'task_update':
      (standardMessage as Partial<TaskUpdateMessage>).task_id = 
        legacyMessage.taskId || legacyMessage.task_id;
      if ('status' in legacyMessage) {
        (standardMessage as Partial<TaskUpdateMessage>).status = legacyMessage.status;
      }
      if ('action' in legacyMessage) {
        (standardMessage as Partial<TaskUpdateMessage>).action = 
          legacyMessage.action as 'create' | 'update' | 'cancel';
      }
      if ('result' in legacyMessage) {
        (standardMessage as Partial<TaskUpdateMessage>).result = legacyMessage.result;
      }
      break;
    
    case 'error':
      if ('errorCode' in legacyMessage || 'error_code' in legacyMessage) {
        (standardMessage as Partial<ErrorMessage>).error_code = 
          legacyMessage.errorCode || legacyMessage.error_code;
      }
      if ('errorDetails' in legacyMessage || 'error_details' in legacyMessage) {
        (standardMessage as Partial<ErrorMessage>).error_details = 
          legacyMessage.errorDetails || legacyMessage.error_details;
      }
      if ('severity' in legacyMessage) {
        (standardMessage as Partial<ErrorMessage>).severity = 
          legacyMessage.severity as 'warning' | 'error' | 'critical';
      }
      if ('recoverable' in legacyMessage) {
        (standardMessage as Partial<ErrorMessage>).recoverable = legacyMessage.recoverable;
      }
      break;
  }
  
  return standardMessage;
}

/**
 * Convert a standardized format message to legacy format.
 * 
 * @param message The standardized message object
 * @returns Legacy format message object
 */
export function standardToLegacy(message: Partial<Message>): any {
  // Create legacy message with camelCase fields
  const legacyMessage: any = {
    id: message.id,
    type: message.type,
    timestamp: message.timestamp,
    sessionId: message.session_id
  };
  
  // Handle content field
  if ('content' in message) {
    legacyMessage.content = message.content;
    if (typeof message.content === 'string') {
      legacyMessage.text = message.content;
    }
  }
  
  // Handle agent vs user specifics
  if (message.from_agent) {
    legacyMessage.fromAgent = message.from_agent;
  }
  if (message.to_agent) {
    legacyMessage.toAgent = message.to_agent;
  }
  if (message.from_user) {
    legacyMessage.fromUser = message.from_user;
  }
  
  // Copy additional fields
  if (message.in_reply_to) {
    legacyMessage.inReplyTo = message.in_reply_to;
  }
  
  if ('streaming' in message) {
    legacyMessage.streaming = message.streaming;
  }
  
  if ('turn_complete' in message) {
    legacyMessage.turnComplete = message.turn_complete;
  }
  
  if (message.metadata) {
    legacyMessage.metadata = message.metadata;
  }
  
  // Handle message type specific fields
  const type = message.type;
  switch (type) {
    case MessageType.CONTEXT_UPDATE:
      const contextMsg = message as Partial<ContextUpdateMessage>;
      if (contextMsg.context_id) {
        legacyMessage.contextId = contextMsg.context_id;
      }
      if (contextMsg.context_data) {
        legacyMessage.contextData = contextMsg.context_data;
      }
      if (contextMsg.target_agents) {
        legacyMessage.targetAgents = contextMsg.target_agents;
      }
      break;
      
    case MessageType.TASK_UPDATE:
      const taskMsg = message as Partial<TaskUpdateMessage>;
      if (taskMsg.task_id) {
        legacyMessage.taskId = taskMsg.task_id;
      }
      if (taskMsg.status) {
        legacyMessage.status = taskMsg.status;
      }
      if (taskMsg.action) {
        legacyMessage.action = taskMsg.action;
      }
      if (taskMsg.result) {
        legacyMessage.result = taskMsg.result;
      }
      break;
      
    case MessageType.ERROR:
      const errorMsg = message as Partial<ErrorMessage>;
      if (errorMsg.error_code) {
        legacyMessage.errorCode = errorMsg.error_code;
      }
      if (errorMsg.error_details) {
        legacyMessage.errorDetails = errorMsg.error_details;
      }
      if (errorMsg.severity) {
        legacyMessage.severity = errorMsg.severity;
      }
      if ('recoverable' in errorMsg) {
        legacyMessage.recoverable = errorMsg.recoverable;
      }
      break;
  }
  
  return legacyMessage;
}

/**
 * Check if a message is in standardized format.
 * 
 * @param message The message to check
 * @returns True if message has standardized format, False otherwise
 */
export function isStandardFormat(message: any): boolean {
  // Primary indicator is snake_case fields
  return 'session_id' in message && 'id' in message;
}

/**
 * Check if a message is in legacy format.
 * 
 * @param message The message to check
 * @returns True if message has legacy format, False otherwise
 */
export function isLegacyFormat(message: any): boolean {
  // Primary indicator is camelCase fields
  return 'sessionId' in message && 'id' in message;
}

/**
 * Normalize a message to standardized format regardless of input format.
 * Adds validation to ensure message has minimum required fields.
 * 
 * @param message Message in either legacy or standardized format
 * @returns Normalized message in standardized format
 * @throws Error if message cannot be normalized and lacks required fields
 */
export function normalizeMessage(message: any): Partial<Message> {
  // Handle null or undefined
  if (!message) {
    throw new Error('Cannot normalize undefined or null message');
  }
  
  // Ensure message is an object
  if (typeof message !== 'object') {
    throw new Error(`Cannot normalize non-object message: ${typeof message}`);
  }

  if (isLegacyFormat(message)) {
    return legacyToStandard(message);
  } else if (isStandardFormat(message)) {
    return message;
  } else {
    // Try to guess format and convert to standard
    if ('sessionId' in message) {
      return legacyToStandard(message);
    } else {
      // Create minimum valid message
      const standard: Partial<BaseMessage> = {
        id: message.id || `msg_${Date.now()}`,
        session_id: message.session_id || 'default',
        type: (message.type as MessageType) || MessageType.TEXT,
        timestamp: message.timestamp || new Date().toISOString()
      };
      
      // Copy any content
      if ('content' in message) {
        standard.content = message.content;
      } else if ('text' in message) {
        standard.content = message.text;
      } else {
        // No content found - set to empty string
        standard.content = '';
      }
      
      // Validate the created message has minimum fields
      if (!standard.id || !standard.session_id || !standard.type) {
        console.warn('Normalized message is missing required fields', standard);
      }
      
      return standard;
    }
  }
}

/**
 * Create a new standardized text message from a user.
 * 
 * @param text The message text content
 * @param sessionId The session ID
 * @param toAgent Optional target agent ID
 * @returns A user text message
 */
export function createUserTextMessage(
  text: string, 
  sessionId: string, 
  toAgent?: string
): UserTextMessage {
  return {
    id: `msg_${Date.now()}_${Math.floor(Math.random() * 1000)}`,
    type: MessageType.TEXT,
    session_id: sessionId,
    from_user: true,
    content: text,
    to_agent: toAgent,
    timestamp: new Date().toISOString()
  };
}

/**
 * Create a new standardized text message from an agent.
 * 
 * @param text The message text content
 * @param sessionId The session ID
 * @param fromAgent The agent ID
 * @param inReplyTo Optional ID of message being replied to
 * @param streaming Whether this is a streaming message
 * @returns An agent text message
 */
export function createAgentTextMessage(
  text: string,
  sessionId: string,
  fromAgent: string,
  inReplyTo?: string,
  streaming: boolean = false
): AgentTextMessage {
  return {
    id: `msg_${Date.now()}_${Math.floor(Math.random() * 1000)}`,
    type: MessageType.TEXT,
    session_id: sessionId,
    from_agent: fromAgent,
    content: text,
    in_reply_to: inReplyTo,
    timestamp: new Date().toISOString(),
    streaming: streaming,
    turn_complete: !streaming
  };
}

File: api_gateway/src/services/message_adapter.py
"""
Message adapter service for TKR Multi-Agent Chat System.

This module provides adapters for converting between standardized and legacy message formats
to ensure compatibility during the transition period.
"""

from datetime import datetime
import uuid
from typing import Dict, Any, List, Optional, Union

from ..models.message_schema import (
    MessageType,
    BaseMessage,
    Message,
    UserTextMessage,
    AgentTextMessage,
    AgentToAgentMessage,
    SystemMessage,
    ErrorMessage,
    ContextUpdateMessage,
    TaskUpdateMessage,
    PingMessage,
    PongMessage
)

def legacy_to_standard(message: Dict[str, Any]) -> Dict[str, Any]:
    """
    Convert a legacy format message to standardized format.
    
    Args:
        message: The legacy message object
        
    Returns:
        Standardized message dictionary
    """
    # Create base message with required fields
    standard_message = {
        "id": message.get("id") or str(uuid.uuid4()),
        "timestamp": message.get("timestamp") or datetime.utcnow().isoformat(),
        "session_id": message.get("sessionId") or message.get("session_id"),
    }
    
    # Convert message type
    msg_type = message.get("type", "text")
    standard_message["type"] = msg_type  # Use same type string, validated elsewhere
    
    # Add content - both content and text are used in legacy messages
    if "content" in message:
        standard_message["content"] = message["content"]
    elif "text" in message:
        standard_message["content"] = message["text"]
    
    # Handle agent vs user specifics
    if message.get("fromAgent") or message.get("from_agent"):
        standard_message["from_agent"] = message.get("fromAgent") or message.get("from_agent")
        # Ensure it's explicitly marked as not a user message
        standard_message["from_user"] = False
    elif message.get("fromUser") or message.get("from_user") or message.get("fromUser") == True or message.get("from_user") == True:
        # Explicitly set to true (handle both boolean and string values)
        standard_message["from_user"] = True
        # Make sure from_agent is not set
        standard_message.pop("from_agent", None)
    else:
        # Default to user message if no source specified
        standard_message["from_user"] = True
        # Make sure from_agent is not set
        standard_message.pop("from_agent", None)
    
    # Handle target agent
    if message.get("toAgent") or message.get("to_agent"):
        standard_message["to_agent"] = message.get("toAgent") or message.get("to_agent")
    
    # Copy additional fields if present
    if "inReplyTo" in message or "in_reply_to" in message:
        standard_message["in_reply_to"] = message.get("inReplyTo") or message.get("in_reply_to")
    
    if "streaming" in message:
        standard_message["streaming"] = message["streaming"]
    
    if "turnComplete" in message or "turn_complete" in message:
        standard_message["turn_complete"] = message.get("turnComplete") or message.get("turn_complete")
    
    if "metadata" in message:
        standard_message["metadata"] = message["metadata"]
    
    # Handle message type specific fields
    if msg_type == "context_update":
        standard_message["context_id"] = message.get("contextId") or message.get("context_id")
        standard_message["context_data"] = message.get("contextData") or message.get("context_data")
        if "targetAgents" in message or "target_agents" in message:
            standard_message["target_agents"] = message.get("targetAgents") or message.get("target_agents")
            
    elif msg_type == "task_update":
        standard_message["task_id"] = message.get("taskId") or message.get("task_id")
        if "status" in message:
            standard_message["status"] = message["status"]
        if "action" in message:
            standard_message["action"] = message["action"]
        if "result" in message:
            standard_message["result"] = message["result"]
            
    elif msg_type == "error":
        if "errorCode" in message or "error_code" in message:
            standard_message["error_code"] = message.get("errorCode") or message.get("error_code")
        if "errorDetails" in message or "error_details" in message:
            standard_message["error_details"] = message.get("errorDetails") or message.get("error_details")
        if "severity" in message:
            standard_message["severity"] = message["severity"]
        if "recoverable" in message:
            standard_message["recoverable"] = message["recoverable"]
    
    return standard_message

def standard_to_legacy(message: Dict[str, Any]) -> Dict[str, Any]:
    """
    Convert a standardized format message to legacy format.
    
    Args:
        message: The standardized message object
        
    Returns:
        Legacy format message dictionary
    """
    # Create legacy message with camelCase fields
    legacy_message = {
        "id": message["id"],
        "type": message["type"],
        "timestamp": message["timestamp"],
        "sessionId": message["session_id"]
    }
    
    # Handle content field
    if "content" in message:
        legacy_message["content"] = message["content"]
        legacy_message["text"] = message["content"] if isinstance(message["content"], str) else str(message["content"])
    
    # Handle agent vs user specifics
    if "from_agent" in message and message["from_agent"]:
        legacy_message["fromAgent"] = message["from_agent"]
    if "to_agent" in message:
        legacy_message["toAgent"] = message["to_agent"]
    if "from_user" in message and message["from_user"]:
        legacy_message["fromUser"] = message["from_user"]
    
    # Copy additional fields
    if "in_reply_to" in message and message["in_reply_to"]:
        legacy_message["inReplyTo"] = message["in_reply_to"]
    
    if "streaming" in message:
        legacy_message["streaming"] = message["streaming"]
    
    if "turn_complete" in message:
        legacy_message["turnComplete"] = message["turn_complete"]
        
    if "metadata" in message and message["metadata"]:
        legacy_message["metadata"] = message["metadata"]
    
    # Handle message type specific fields
    msg_type = message["type"]
    if msg_type == "context_update":
        if "context_id" in message:
            legacy_message["contextId"] = message["context_id"]
        if "context_data" in message:
            legacy_message["contextData"] = message["context_data"]
        if "target_agents" in message:
            legacy_message["targetAgents"] = message["target_agents"]
            
    elif msg_type == "task_update":
        if "task_id" in message:
            legacy_message["taskId"] = message["task_id"]
        if "status" in message:
            legacy_message["status"] = message["status"]
        if "action" in message:
            legacy_message["action"] = message["action"]
        if "result" in message:
            legacy_message["result"] = message["result"]
            
    elif msg_type == "error":
        if "error_code" in message:
            legacy_message["errorCode"] = message["error_code"]
        if "error_details" in message:
            legacy_message["errorDetails"] = message["error_details"]
        if "severity" in message:
            legacy_message["severity"] = message["severity"]
        if "recoverable" in message:
            legacy_message["recoverable"] = message["recoverable"]
    
    return legacy_message

def is_standard_format(message: Dict[str, Any]) -> bool:
    """
    Check if a message is in standardized format.
    
    Args:
        message: The message to check
        
    Returns:
        True if message has standardized format, False otherwise
    """
    # Primary indicator is snake_case fields
    return "session_id" in message and "id" in message

def is_legacy_format(message: Dict[str, Any]) -> bool:
    """
    Check if a message is in legacy format.
    
    Args:
        message: The message to check
        
    Returns:
        True if message has legacy format, False otherwise
    """
    # Primary indicator is camelCase fields
    return "sessionId" in message and "id" in message

def normalize_message(message: Dict[str, Any]) -> Dict[str, Any]:
    """
    Normalize a message to standardized format regardless of input format.
    Adds validation to ensure message has minimum required fields.
    
    Args:
        message: Message in either legacy or standardized format
        
    Returns:
        Normalized message in standardized format
        
    Raises:
        ValueError: If message cannot be normalized or is missing required fields
        TypeError: If message is not a dictionary
    """
    # Handle None or empty dictionaries
    if message is None:
        raise ValueError("Cannot normalize None message")
    
    # Ensure message is a dictionary
    if not isinstance(message, dict):
        raise TypeError(f"Message must be a dictionary, got {type(message)}")
    
    if not message:
        raise ValueError("Cannot normalize empty message dictionary")
    
    if is_legacy_format(message):
        return legacy_to_standard(message)
    elif is_standard_format(message):
        return message
    else:
        # Try to guess format and convert to standard
        if "sessionId" in message:
            return legacy_to_standard(message)
        else:
            # Create minimum valid message
            standard = {
                "id": message.get("id") or str(uuid.uuid4()),
                "session_id": message.get("session_id") or "default",
                "type": message.get("type") or "text",
                "timestamp": message.get("timestamp") or datetime.utcnow().isoformat()
            }
            
            # Copy any content
            if "content" in message:
                standard["content"] = message["content"]
            elif "text" in message:
                standard["content"] = message["text"]
            else:
                # No content found - set to empty string
                standard["content"] = ""
            
            # Validate the created message has minimum fields
            if not standard.get("id") or not standard.get("session_id") or not standard.get("type"):
                import logging
                logging.getLogger(__name__).warning(f"Normalized message is missing required fields: {standard}")
            
            return standard